File: event-dashboard/src/utils/round4Scoring.ts
import {
  Round4StageId,
  StageTimingRecord,
  RubricCategoryConfig,
  FinalScoreFormulaConfig,
  Round4Config,
  TeamRound4Record,
  Round4EngineResult,
  Round4SummaryStats,
  TeamPair,
  JudgeScoreRecord,
  AgentGuessingRecord,
  TeamFinalScoreBreakdown,
  Round4ChecklistItem,
} from '../types/round4';

export const DEFAULT_RUBRIC_CATEGORIES: RubricCategoryConfig[] = [
  {
    id: 'logical_structure',
    name: 'Logical structure',
    maxMarks: 20, // Suggested — pending organizer confirmation
    isConfirmed: false,
    description: 'Coherence of legal reasoning, clarity of core premise, structured argumentation',
  },
  {
    id: 'evidence_use',
    name: 'Use of evidence',
    maxMarks: 20, // Suggested — pending organizer confirmation
    isConfirmed: false,
    description: 'Application of case facts, documentary evidence, and discovery filings',
  },
  {
    id: 'rebuttal',
    name: 'Rebuttal',
    maxMarks: 20, // Suggested — pending organizer confirmation
    isConfirmed: false,
    description: 'Responsiveness to opposing counsel, counter-argument sharpness, cross-examination',
  },
  {
    id: 'resource_questioning',
    name: 'Questioning the resource person',
    maxMarks: 15, // Suggested — pending organizer confirmation
    isConfirmed: false,
    description: 'Relevance of queries, strategic discovery extraction, professional inquiry',
  },
  {
    id: 'presentation_teamwork',
    name: 'Presentation and teamwork',
    maxMarks: 15, // Suggested — pending organizer confirmation
    isConfirmed: false,
    description: 'Oratorical delivery, courtroom decorum, equal contribution across squad members',
  },
  {
    id: 'time_management',
    name: 'Time management',
    maxMarks: 10, // Suggested — pending organizer confirmation
    isConfirmed: false,
    description: 'Adherence to oral argument time limits, efficient division of speaking segments',
  },
];

export function createInitialPairStages(): Record<Round4StageId, StageTimingRecord> {
  return {
    prep_1: {
      stageId: 'prep_1',
      name: 'Preparation 1',
      suggestedDurationMinutes: 40,
      configuredDurationMinutes: 40,
      status: 'not_started',
      startedAt: null,
      endedAt: null,
      actualDurationSeconds: null,
      notes: '',
      incidentFlags: '',
    },
    hearing_1: {
      stageId: 'hearing_1',
      name: 'Hearing 1',
      suggestedDurationMinutes: 20,
      configuredDurationMinutes: 20,
      status: 'not_started',
      startedAt: null,
      endedAt: null,
      actualDurationSeconds: null,
      notes: '',
      incidentFlags: '',
    },
    file_exchange: {
      stageId: 'file_exchange',
      name: 'Opposing-File Exchange',
      suggestedDurationMinutes: null, // duration not specified in documentation
      configuredDurationMinutes: null,
      status: 'not_started',
      startedAt: null,
      endedAt: null,
      actualDurationSeconds: null,
      notes: '',
      incidentFlags: '',
    },
    prep_2: {
      stageId: 'prep_2',
      name: 'Preparation 2',
      suggestedDurationMinutes: 25, // 20-25 minutes suggested
      configuredDurationMinutes: 25,
      status: 'not_started',
      startedAt: null,
      endedAt: null,
      actualDurationSeconds: null,
      notes: '',
      incidentFlags: '',
    },
    hearing_2: {
      stageId: 'hearing_2',
      name: 'Hearing 2',
      suggestedDurationMinutes: 20,
      configuredDurationMinutes: 20,
      status: 'not_started',
      startedAt: null,
      endedAt: null,
      actualDurationSeconds: null,
      notes: '',
      incidentFlags: '',
    },
  };
}

export const DEFAULT_FINAL_SCORE_FORMULA: FinalScoreFormulaConfig = {
  // Suggested formula: Final Score = Legal Battle panel score + Agent guessing points + 10% of remaining Black Market points
  // Organizers may adjust weighting. Labeled "Suggested — pending organizer confirmation".
  panelScoreWeight: 1.0,
  agentGuessingWeight: 1.0,
  blackMarketWeightPercent: 10, // 10%
  isFormulaConfirmed: false,
  confirmedAt: null,
  confirmedBy: null,
};

export const DEFAULT_ROUND4_CONFIG: Round4Config = {
  rubricCategories: DEFAULT_RUBRIC_CATEGORIES,
  isRubricConfirmed: false, // Suggested — pending organizer confirmation
  judgeAggregation: 'average',
  judgesList: [
    { id: 'judge-1', name: 'Faculty Judge 1 [TBD]' },
    { id: 'judge-2', name: 'Faculty Judge 2 [TBD]' },
  ],
  isGuessingRulesConfigured: false,
  guessingPointsForCorrect: null, // Unconfigured
  guessingPointsForIncorrect: null,
  finalScoreFormula: DEFAULT_FINAL_SCORE_FORMULA,
  advancingTeamsCount: null, // Configurable: null = unconfirmed (e.g. Top 3 for Grand Finale podium)
  isFinalized: false,
  finalizedAt: null,
};

/**
 * Calculates a squad's aggregated panel score from submitted judge scorecards.
 * Never treats missing scores as zero.
 */
export function calculatePanelScore(
  judgeScores: JudgeScoreRecord[],
  aggregation: 'average' | 'sum' | 'single_judge'
): { panelScore: number | null; isComplete: boolean } {
  const submittedScores = judgeScores.filter((js) => js.isSubmitted);
  if (submittedScores.length === 0) {
    return { panelScore: null, isComplete: false };
  }

  if (aggregation === 'single_judge') {
    return { panelScore: submittedScores[0].totalScore, isComplete: true };
  }

  if (aggregation === 'sum') {
    const sum = submittedScores.reduce((acc, s) => acc + s.totalScore, 0);
    return { panelScore: sum, isComplete: true };
  }

  // Default: average
  const sum = submittedScores.reduce((acc, s) => acc + s.totalScore, 0);
  const avg = Number((sum / submittedScores.length).toFixed(2));
  return { panelScore: avg, isComplete: true };
}

/**
 * Calculates a squad's final score using the configurable formula.
 *
 * Suggested formula:
 * Final Score = (Panel Score * weight) + (Agent Guessing Points * weight) + (Black Market Remaining Points * percentage)
 *
 * Enforces:
 * - Does NOT silently treat missing scores as zero.
 * - Does NOT calculate final score if formula is unconfirmed by organizers.
 */
export function calculateFinalScoreBreakdown(
  teamId: string,
  panelScore: number | null,
  agentRecord: AgentGuessingRecord | undefined,
  blackMarketBalance: number,
  formula: FinalScoreFormulaConfig,
  isGuessingConfigured: boolean
): TeamFinalScoreBreakdown {
  const missingComponents: string[] = [];

  if (!formula.isFormulaConfirmed) {
    missingComponents.push('Final score formula unconfirmed by organizers');
  }

  if (panelScore === null) {
    missingComponents.push('Faculty judging panel score missing');
  }

  let agentGuessingPoints: number | null = null;
  if (isGuessingConfigured) {
    if (!agentRecord || !agentRecord.isVerified || agentRecord.pointsAwarded === null) {
      missingComponents.push('Secret agent guessing score unverified');
    } else {
      agentGuessingPoints = agentRecord.pointsAwarded;
    }
  } else {
    // If guessing rules not configured, record as null
    agentGuessingPoints = agentRecord?.pointsAwarded ?? null;
  }

  const rawPanel = panelScore;
  const weightedPanel = panelScore !== null ? Number((panelScore * formula.panelScoreWeight).toFixed(2)) : null;
  const bmContribution = Number(
    ((Math.max(0, blackMarketBalance) * formula.blackMarketWeightPercent) / 100).toFixed(2)
  );

  let finalScore: number | null = null;
  const isComplete = missingComponents.length === 0 && panelScore !== null;

  if (isComplete) {
    let sum = (weightedPanel ?? 0) + bmContribution;
    if (agentGuessingPoints !== null) {
      sum += agentGuessingPoints * formula.agentGuessingWeight;
    }
    finalScore = Number(sum.toFixed(2));
  }

  return {
    teamId,
    rawPanelScore: rawPanel,
    weightedPanelScore: weightedPanel,
    agentGuessingPoints,
    blackMarketBalance,
    blackMarketContribution: bmContribution,
    finalScore,
    isComplete,
    missingComponents,
  };
}

/**
 * Standings and Qualification Engine for Round 4: The Legal Battle.
 *
 * Evaluates:
 * - Verification that exactly 8 teams qualified from Round 3.
 * - Confirmation of team pairings and case assignments.
 * - Stage progression across Preparation 1, Hearing 1, File Exchange, Prep 2, Hearing 2.
 * - Faculty judging panel score completion.
 * - Final score formula confirmation and calculation.
 * - Tie detection across onward qualification cutoff.
 */
export function processRound4Standings(
  records: TeamRound4Record[],
  pairs: TeamPair[],
  config: Round4Config,
  round3Finalized: boolean
): Round4EngineResult {
  const checklist: Round4ChecklistItem[] = [];

  // 1. Check Round 3 finalization
  const r3Check = round3Finalized;
  checklist.push({
    id: 'round-3-finalized',
    label: 'Round 3: The Black Market Finalized',
    passed: r3Check,
    details: r3Check
      ? 'Official Round 3 results sealed. 8 finalist squads advanced.'
      : 'Round 3 results must be finalized before Round 4 can proceed.',
    severity: 'blocker',
  });

  // 2. Check exactly 8 participating squads
  const teamCountCheck = records.length === 8;
  checklist.push({
    id: 'squad-count',
    label: 'Exactly 8 Finalist Squads Participating',
    passed: teamCountCheck,
    details: `Detected ${records.length} participating squads (Expected: 8).`,
    severity: 'blocker',
  });

  // 3. Check pairings confirmation
  const pairsFormed = pairs.length === 4 && pairs.every((p) => p.teamAId && p.teamBId);
  const pairingsConfirmed = pairsFormed && pairs.every((p) => p.isConfirmed);
  checklist.push({
    id: 'pairings-confirmed',
    label: '4 Team Matchup Pairings Confirmed',
    passed: pairingsConfirmed,
    details: pairingsConfirmed
      ? 'All 4 head-to-head pairings officially confirmed by organizers.'
      : 'Pairings must be formed and explicitly confirmed by organizers.',
    severity: 'blocker',
  });

  // 4. Check case assignments and sides
  const casesAssigned =
    pairs.length === 4 &&
    pairs.every(
      (p) =>
        p.caseName &&
        p.caseName.trim().length > 0 &&
        p.teamAAssignment.side !== 'Unassigned' &&
        p.teamBAssignment.side !== 'Unassigned'
    );
  checklist.push({
    id: 'cases-assigned',
    label: 'Fictional Legal Cases & Sides Assigned',
    passed: casesAssigned,
    details: casesAssigned
      ? 'All 4 pairs have designated legal cases and counsel sides.'
      : 'Assign official case titles and team sides for all 4 pairs.',
    severity: 'blocker',
  });

  // 5. Check stage progression (hearings and file exchange)
  const hearingsDone = pairs.every(
    (p) =>
      p.stages.hearing_1.status === 'completed' &&
      p.stages.file_exchange.status === 'completed' &&
      p.stages.hearing_2.status === 'completed'
  );
  checklist.push({
    id: 'hearings-completed',
    label: 'Hearing Stages & File Exchange Completed',
    passed: hearingsDone,
    details: hearingsDone
      ? 'Hearing 1, Opposing-File Exchange, and Hearing 2 logged complete.'
      : 'Oral argument sessions and file exchanges must be marked completed.',
    severity: 'blocker',
  });

  // 6. Check faculty judging completeness
  const judgesComplete = records.every((r) => r.isJudgePanelComplete && r.panelScore !== null);
  checklist.push({
    id: 'judging-scores',
    label: 'Faculty Panel Scorecards Submitted',
    passed: judgesComplete,
    details: judgesComplete
      ? 'All 8 finalist squads have complete submitted judging scores.'
      : 'Faculty judging scorecards are missing or incomplete for one or more squads.',
    severity: 'blocker',
  });

  // 7. Check final score formula confirmation
  const formulaConfirmed = config.finalScoreFormula.isFormulaConfirmed;
  checklist.push({
    id: 'formula-confirmed',
    label: 'Final-Score Formula Officially Confirmed',
    passed: formulaConfirmed,
    details: formulaConfirmed
      ? 'Scoring formula and weighting confirmed by organizing committee.'
      : 'Suggested scoring formula is unconfirmed. Organizers must review and confirm.',
    severity: 'blocker',
  });

  // Sort records:
  // If final scores are available, sort by finalScore descending.
  // Otherwise sort by panelScore descending or teamNumber.
  const sortedRecords = [...records].sort((a, b) => {
    const scoreA = a.finalScoreBreakdown.finalScore;
    const scoreB = b.finalScoreBreakdown.finalScore;

    if (scoreA !== null && scoreB !== null) {
      if (scoreB !== scoreA) return scoreB - scoreA;
    } else if (scoreA !== null) {
      return -1;
    } else if (scoreB !== null) {
      return 1;
    } else {
      // Fallback to panel score
      const pA = a.panelScore ?? -1;
      const pB = b.panelScore ?? -1;
      if (pB !== pA) return pB - pA;
    }

    return a.teamNumber - b.teamNumber;
  });

  // Assign ranks if final scores are computed
  let currentRank = 1;
  sortedRecords.forEach((rec) => {
    if (rec.finalScoreBreakdown.finalScore !== null) {
      rec.rank = currentRank++;
    } else {
      rec.rank = null;
    }
  });

  // Detect ties across advancing cutoff if onward qualification count is configured
  let tiesAffectingCutoff = false;
  if (config.advancingTeamsCount && config.advancingTeamsCount > 0 && config.advancingTeamsCount < 8) {
    const cutoffRank = config.advancingTeamsCount;
    const scoreGroups = new Map<number, TeamRound4Record[]>();

    sortedRecords.forEach((rec) => {
      const fs = rec.finalScoreBreakdown.finalScore;
      if (fs !== null) {
        if (!scoreGroups.has(fs)) scoreGroups.set(fs, []);
        scoreGroups.get(fs)!.push(rec);
      }
    });

    scoreGroups.forEach((group, scoreVal) => {
      if (group.length > 1) {
        const ranks = group.map((r) => r.rank!).filter((r) => r !== null && r !== undefined);
        const minRank = Math.min(...ranks);
        const maxRank = Math.max(...ranks);

        if (minRank <= cutoffRank && maxRank > cutoffRank) {
          tiesAffectingCutoff = true;
          group.forEach((rec) => {
            rec.tieRequiresReview = true;
            rec.tieReason = `Tied on ${scoreVal} pts spanning advancing cutoff #${cutoffRank} (Ranks #${minRank}–#${maxRank}). Official tournament rules require manual marshal review.`;
          });
        }
      }
    });
  }

  checklist.push({
    id: 'tie-safeguard',
    label: 'Advancement Cutoff Free of Unresolved Ties',
    passed: !tiesAffectingCutoff,
    details: tiesAffectingCutoff
      ? 'An unresolved tie spans across the onward qualification cutoff boundary.'
      : 'No cutoff-affecting ties detected.',
    severity: 'blocker',
  });

  // Assign review status
  sortedRecords.forEach((rec) => {
    if (!round3Finalized) {
      rec.reviewStatus = 'Round 3 Pending';
    } else if (config.isFinalized) {
      if (config.advancingTeamsCount && rec.rank !== null && rec.rank !== undefined && rec.rank <= config.advancingTeamsCount) {
        rec.reviewStatus = 'Finalized Qualified';
      } else {
        rec.reviewStatus = 'Finalized Eliminated';
      }
    } else if (rec.tieRequiresReview) {
      rec.reviewStatus = 'Tie Review Needed';
    } else if (!pairingsConfirmed) {
      rec.reviewStatus = 'Pairing Unconfirmed';
    } else if (rec.side === 'Unassigned' || !rec.caseName) {
      rec.reviewStatus = 'Case Unassigned';
    } else if (rec.stagesCompletedCount < 3) {
      rec.reviewStatus = 'Hearing Stage';
    } else if (!rec.isJudgePanelComplete) {
      rec.reviewStatus = 'Awaiting Scores';
    } else if (!formulaConfirmed) {
      rec.reviewStatus = 'Formula Unconfirmed';
    } else {
      rec.reviewStatus = 'Ready for Review';
    }
  });

  // Evaluate canFinalize
  const blockerIssues = checklist.filter((c) => !c.passed && c.severity === 'blocker');
  const canFinalize = blockerIssues.length === 0;
  const blockReason = blockerIssues.length > 0 ? blockerIssues[0].label + ': ' + blockerIssues[0].details : null;

  const advancingCount = config.advancingTeamsCount ?? 0;
  const advancingTeamIds = sortedRecords
    .filter((r) => r.rank !== null && r.rank !== undefined && advancingCount > 0 && r.rank <= advancingCount)
    .map((r) => r.teamId);

  const eliminatedTeamIds = sortedRecords
    .filter((r) => r.rank !== null && r.rank !== undefined && advancingCount > 0 && r.rank > advancingCount)
    .map((r) => r.teamId);

  return {
    records: sortedRecords,
    canFinalize,
    blockReason,
    checklist,
    tiesAffectingCutoff,
    advancingTeamIds,
    eliminatedTeamIds,
  };
}

/**
 * Computes summary KPI statistics for Round 4.
 */
export function computeRound4SummaryStats(
  records: TeamRound4Record[],
  pairs: TeamPair[],
  config: Round4Config,
  round3Finalized: boolean,
  round3QualifiedCount: number,
  engine: Round4EngineResult
): Round4SummaryStats {
  const pairingsConfirmed = pairs.length === 4 && pairs.every((p) => p.isConfirmed);
  const casesAssignedCount = pairs.filter((p) => p.caseName && p.caseName.trim().length > 0).length;

  let prep1 = 0;
  let hearing1 = 0;
  let exchange = 0;
  let prep2 = 0;
  let hearing2 = 0;

  pairs.forEach((p) => {
    if (p.stages.prep_1.status === 'completed') prep1++;
    if (p.stages.hearing_1.status === 'completed') hearing1++;
    if (p.stages.file_exchange.status === 'completed') exchange++;
    if (p.stages.prep_2.status === 'completed') prep2++;
    if (p.stages.hearing_2.status === 'completed') hearing2++;
  });

  const judgingCompletedCount = records.filter((r) => r.isJudgePanelComplete).length;
  const agentGuessesVerifiedCount = records.filter(
    (r) => r.agentGuessingRecord && r.agentGuessingRecord.isVerified
  ).length;

  return {
    round3Finalized,
    eligibleTeamsCount: round3QualifiedCount,
    pairsConfiguredCount: pairs.length,
    pairingsConfirmed,
    casesAssignedCount,
    prep1CompletedCount: prep1,
    hearing1CompletedCount: hearing1,
    fileExchangeCompletedCount: exchange,
    prep2CompletedCount: prep2,
    hearing2CompletedCount: hearing2,
    judgingCompletedCount,
    agentGuessesVerifiedCount,
    scoringFormulaConfirmed: config.finalScoreFormula.isFormulaConfirmed,
    isRubricConfirmed: config.isRubricConfirmed,
    canFinalize: engine.canFinalize,
    isFinalized: config.isFinalized,
  };
}

File: backend/app/scoring/round4_scoring.py
from typing import List, Dict, Any, Optional
from app.core.constants import (
    R4_FINALISTS,
    R4_ADVANCING_COUNT,
    R4_RUBRIC_LOGICAL_STRUCTURE_MAX,
    R4_RUBRIC_EVIDENCE_MAX,
    R4_RUBRIC_REBUTTAL_MAX,
    R4_RUBRIC_RESOURCE_PERSON_MAX,
    R4_RUBRIC_PRESENTATION_TEAMWORK_MAX,
    R4_RUBRIC_TIME_MAX,
    R4_RUBRIC_TOTAL_MAX,
    R4_RUBRIC_LIMITS,
)

# Canonical rubric criteria keys
RUBRIC_CANONICAL_KEYS = [
    "logical_structure",
    "evidence",
    "rebuttal",
    "resource_person_questioning",
    "presentation_teamwork",
    "time",
]

RUBRIC_KEY_ALIASES = {
    "logical_structure": "logical_structure",
    "evidence": "evidence",
    "evidence_use": "evidence",
    "rebuttal": "rebuttal",
    "resource_person_questioning": "resource_person_questioning",
    "resource_questioning": "resource_person_questioning",
    "presentation_teamwork": "presentation_teamwork",
    "time": "time",
    "time_management": "time",
}

DEFAULT_CATEGORY_LIMITS = {
    "logical_structure": R4_RUBRIC_LOGICAL_STRUCTURE_MAX,  # 20.0
    "evidence": R4_RUBRIC_EVIDENCE_MAX,                    # 20.0
    "evidence_use": R4_RUBRIC_EVIDENCE_MAX,                # 20.0
    "rebuttal": R4_RUBRIC_REBUTTAL_MAX,                    # 20.0
    "resource_person_questioning": R4_RUBRIC_RESOURCE_PERSON_MAX,  # 15.0
    "resource_questioning": R4_RUBRIC_RESOURCE_PERSON_MAX,          # 15.0
    "presentation_teamwork": R4_RUBRIC_PRESENTATION_TEAMWORK_MAX,  # 15.0
    "time": R4_RUBRIC_TIME_MAX,                            # 10.0
    "time_management": R4_RUBRIC_TIME_MAX,                 # 10.0
}

LEGACY_CATEGORY_LIMITS = {
    "arguments": 30.0,
    "crossExam": 25.0,
    "cross_exam": 25.0,
    "evidence": 25.0,
    "demeanor": 20.0,
}


def normalize_rubric_key(key: str) -> str:
    """Normalize alias keys to canonical rubric keys."""
    return RUBRIC_KEY_ALIASES.get(key, key)


def validate_rubric_scores(
    scores: Dict[str, float],
    rubric_categories: Optional[List[Dict[str, Any]]] = None
) -> Dict[str, float]:
    """
    Validates that:
    1. Score values are non-negative.
    2. Category marks do not exceed defined maximums for defined criteria.
    3. Total score does not exceed 100.0 (or custom rubric maximum).
    """
    if not scores:
        raise ValueError("Judge scorecard scores cannot be empty.")

    is_legacy_format = any(k in ("arguments", "crossExam", "cross_exam", "demeanor") for k in scores)

    known_limits: Dict[str, float] = {}
    if is_legacy_format:
        known_limits.update(LEGACY_CATEGORY_LIMITS)
    elif rubric_categories:
        for cat in rubric_categories:
            cat_id = cat.get("id")
            if cat_id:
                max_m = float(cat.get("maxMarks", 20.0))
                known_limits[cat_id] = max_m
                norm_k = normalize_rubric_key(cat_id)
                known_limits[norm_k] = max_m
    else:
        for k, v in DEFAULT_CATEGORY_LIMITS.items():
            known_limits[k] = v

    normalized: Dict[str, float] = {}
    total = 0.0

    for raw_key, raw_val in scores.items():
        try:
            val = float(raw_val)
        except (TypeError, ValueError):
            raise ValueError(f"Score for '{raw_key}' must be a numeric value, got '{raw_val}'.")

        if val < 0.0:
            raise ValueError(f"Score for '{raw_key}' cannot be negative (got {val}).")

        # Check limit if key is in known_limits
        max_allowed = known_limits.get(raw_key)
        if max_allowed is None:
            norm_k = normalize_rubric_key(raw_key)
            if norm_k in known_limits:
                max_allowed = known_limits[norm_k]

        if max_allowed is not None and val > max_allowed:
            raise ValueError(
                f"Score {val} for '{raw_key}' exceeds maximum allowed marks of {max_allowed}."
            )
        elif max_allowed is None and val > R4_RUBRIC_TOTAL_MAX:
            raise ValueError(
                f"Score {val} for '{raw_key}' exceeds maximum rubric limit of {R4_RUBRIC_TOTAL_MAX}."
            )

        normalized[raw_key] = val
        total += val

    max_total = R4_RUBRIC_TOTAL_MAX
    if total > max_total:
        raise ValueError(f"Total score {round(total, 2)} exceeds maximum rubric limit of {max_total}.")

    return normalized


def calculate_legal_battle_score(
    logical_structure: float,
    evidence: float,
    rebuttal: float,
    resource_person_questioning: float,
    presentation_teamwork: float,
    time: float
) -> float:
    """
    Direct calculation function accepting the 6 official rubric criteria out of 100.
    """
    scores = {
        "logical_structure": logical_structure,
        "evidence": evidence,
        "rebuttal": rebuttal,
        "resource_person_questioning": resource_person_questioning,
        "presentation_teamwork": presentation_teamwork,
        "time": time,
    }
    validated = validate_rubric_scores(scores)
    return round(sum(validated.values()), 2)


def calculate_panel_score(
    judge_scores: List[Dict[str, Any]],
    aggregation: str = "average"
) -> Dict[str, Any]:
    """
    Aggregates multi-judge scorecards for a squad.
    """
    submitted = [s for s in judge_scores if s.get("is_submitted")]
    if not submitted:
        return {"panel_score": None, "is_complete": False, "submitted_count": 0}

    scores = [float(s["total_score"]) for s in submitted]

    if aggregation == "single_judge":
        return {"panel_score": scores[0], "is_complete": True, "submitted_count": len(scores)}
    elif aggregation == "sum":
        return {"panel_score": round(sum(scores), 2), "is_complete": True, "submitted_count": len(scores)}
    elif aggregation == "median":
        sorted_s = sorted(scores)
        n = len(sorted_s)
        med = sorted_s[n // 2] if n % 2 != 0 else (sorted_s[n // 2 - 1] + sorted_s[n // 2]) / 2.0
        return {"panel_score": round(med, 2), "is_complete": True, "submitted_count": len(scores)}
    else:  # default: average
        avg = round(sum(scores) / len(scores), 2)
        return {"panel_score": avg, "is_complete": True, "submitted_count": len(scores)}


def calculate_final_score_breakdown(
    team_id: str,
    panel_score: Optional[float],
    agent_record: Optional[Dict[str, Any]] = None,
    black_market_balance: float = 0.0,
    formula: Optional[Dict[str, Any]] = None,
    is_guessing_configured: bool = False
) -> Dict[str, Any]:
    """
    Calculates the team's official Round 4 Legal Battle score breakdown.

    OFFICIAL EVENT RULES:
    1. The Round 4 Legal Battle score is STRICTLY based on the official 100-point Legal Battle rubric:
       - Logical Structure: 20 max
       - Evidence: 20 max
       - Rebuttal: 20 max
       - Resource Person Questioning: 15 max
       - Presentation / Teamwork: 15 max
       - Time: 10 max
       - TOTAL: 100 max
    2. Secret Agent guessing bonuses/penalties and Black Market balances are NOT included in the
       Round 4 Legal Battle score (Secret Agent guessing is officially scored in the Finale / Step 14).
    3. The team's score equals their pure panel score out of 100.
    """
    missing_components = []
    if panel_score is None:
        missing_components.append("Faculty judging panel score missing")

    is_complete = (panel_score is not None)
    final_score = round(panel_score, 2) if panel_score is not None else None

    return {
        "team_id": team_id,
        "raw_panel_score": panel_score,
        "weighted_panel_score": panel_score,
        "agent_guessing_points": None,  # Not part of Round 4 Legal Battle score
        "black_market_balance": black_market_balance,
        "black_market_contribution": 0.0,  # Not added to Round 4 Legal Battle score
        "final_score": final_score,
        "is_complete": is_complete,
        "missing_components": missing_components
    }


def process_round4_standings(
    records: List[Dict[str, Any]],
    pairs: List[Dict[str, Any]],
    config: Dict[str, Any],
    round3_finalized: bool
) -> Dict[str, Any]:
    checklist = []
    issues = []

    # 1. Round 3 finalized
    checklist.append({
        "id": "round-3-finalized",
        "label": "Round 3 Finalized",
        "passed": round3_finalized,
        "details": "Round 3 results must be finalized before Round 4 proceeds."
    })
    if not round3_finalized:
        issues.append({"code": "PREVIOUS_ROUND_UNFINALIZED", "message": "Round 3 results must be finalized before Round 4 proceeds."})

    # 2. Squad count == 8
    count_ok = len(records) == 8
    checklist.append({
        "id": "squad-count",
        "label": "Exactly 8 Finalist Squads Participating",
        "passed": count_ok,
        "details": f"Found {len(records)} squads (Expected: 8)."
    })
    if not count_ok:
        issues.append({"code": "INVALID_TEAM_COUNT", "message": f"Expected exactly 8 participating squads, found {len(records)}."})

    # 3. Pairings confirmed (4 pairs)
    pairs_ok = len(pairs) == 4 and all(p.get("team_a_id") and p.get("team_b_id") and p.get("is_confirmed") for p in pairs)
    checklist.append({
        "id": "pairings-confirmed",
        "label": "4 Team Matchup Pairings Confirmed",
        "passed": pairs_ok,
        "details": "All 4 head-to-head pairings must be confirmed."
    })
    if not pairs_ok:
        issues.append({"code": "PAIRINGS_UNCONFIRMED", "message": "All 4 head-to-head pairings must be created and confirmed by organizers."})

    # 4. Cases assigned
    cases_ok = len(pairs) == 4 and all(
        p.get("case_name") and
        p.get("team_a_side", "Unassigned") != "Unassigned" and
        p.get("team_b_side", "Unassigned") != "Unassigned"
        for p in pairs
    )
    checklist.append({
        "id": "cases-assigned",
        "label": "Fictional Legal Cases & Sides Assigned",
        "passed": cases_ok,
        "details": "Case names and team sides must be assigned for all pairs."
    })
    if not cases_ok:
        issues.append({"code": "CASES_UNASSIGNED", "message": "Case names and team sides must be assigned for all 4 pairs."})

    # 5. Stages completed
    def stages_done(p):
        stages = p.get("stages", {})
        h1 = stages.get("hearing_1", {}).get("status") == "completed"
        fe = stages.get("file_exchange", {}).get("status") == "completed"
        h2 = stages.get("hearing_2", {}).get("status") == "completed"
        return h1 and fe and h2

    hearings_ok = len(pairs) == 4 and all(stages_done(p) for p in pairs)
    checklist.append({
        "id": "hearings-completed",
        "label": "Hearing Stages & File Exchange Completed",
        "passed": hearings_ok,
        "details": "Hearings and file exchange must be logged as completed."
    })
    if not hearings_ok:
        issues.append({"code": "HEARINGS_INCOMPLETE", "message": "Hearing stages and file exchanges must be completed for all matchups."})

    # 6. Faculty judging complete
    judging_ok = len(records) == 8 and all(r.get("is_judge_panel_complete") and r.get("panel_score") is not None for r in records)
    checklist.append({
        "id": "judging-scores",
        "label": "Faculty Panel Scorecards Submitted",
        "passed": judging_ok,
        "details": "All finalist squads must have submitted judging scores."
    })
    if not judging_ok:
        issues.append({"code": "JUDGING_INCOMPLETE", "message": "Faculty judging scorecards are missing or incomplete."})

    # 7. Rubric confirmation
    rubric_confirmed = bool(config.get("is_rubric_confirmed", True))
    checklist.append({
        "id": "rubric-confirmed",
        "label": "Legal Battle Rubric Confirmed",
        "passed": rubric_confirmed,
        "details": "Official 100-point Legal Battle rubric confirmed."
    })

    # Sort records strictly by pure panel_score descending, then team_number
    def sort_key(r):
        ps = r.get("panel_score")
        ps_val = ps if ps is not None else -999999.0
        return (-ps_val, r.get("team_number", 999))

    sorted_records = list(records)
    sorted_records.sort(key=sort_key)

    current_rank = 1
    for r in sorted_records:
        if r.get("final_score_breakdown", {}).get("final_score") is not None:
            r["rank"] = current_rank
            current_rank += 1
        else:
            r["rank"] = None

    # CRITICAL: Under official rules, advancing_teams_count defaults to 8 (NO ELIMINATION)
    advancing_count = config.get("advancing_teams_count", R4_ADVANCING_COUNT)
    if advancing_count is None:
        advancing_count = R4_ADVANCING_COUNT

    ties_affecting_cutoff = False
    tied_teams_at_cutoff = []

    if 0 < advancing_count < len(sorted_records):
        score_groups = {}
        for r in sorted_records:
            fs = r.get("final_score_breakdown", {}).get("final_score")
            if fs is not None:
                score_groups.setdefault(fs, []).append(r)

        for fs, group in score_groups.items():
            if len(group) > 1:
                ranks = [r["rank"] for r in group if r.get("rank") is not None]
                if ranks:
                    min_rank = min(ranks)
                    max_rank = max(ranks)
                    if min_rank <= advancing_count and max_rank > advancing_count:
                        ties_affecting_cutoff = True
                        for r in group:
                            r["tie_requires_review"] = True
                            r["tie_reason"] = f"Tied on {fs} pts across advancing cutoff #{advancing_count}. Manual review required."
                            tied_teams_at_cutoff.append(r["team_id"])

    checklist.append({
        "id": "tie-safeguard",
        "label": "Advancement Cutoff Free of Unresolved Ties",
        "passed": not ties_affecting_cutoff,
        "details": "No cutoff-affecting ties detected." if not ties_affecting_cutoff else "Cutoff tie detected."
    })
    if ties_affecting_cutoff:
        issues.append({"code": "CUTOFF_TIE", "message": f"An unresolved tie spans across advancing cutoff #{advancing_count}. Manual review required."})

    can_finalize = len(issues) == 0
    block_reason = issues[0]["message"] if issues else None

    # Mark advancing status for each record
    for r in sorted_records:
        r["is_advancing"] = bool(r.get("rank") and r["rank"] <= advancing_count)

    advancing_team_ids = [r["team_id"] for r in sorted_records if r.get("rank") and r["rank"] <= advancing_count]
    eliminated_team_ids = [r["team_id"] for r in sorted_records if r.get("rank") and r["rank"] > advancing_count]

    return {
        "records": sorted_records,
        "can_finalize": can_finalize,
        "issues": issues,
        "block_reason": block_reason,
        "checklist": checklist,
        "ties_affecting_cutoff": ties_affecting_cutoff,
        "tied_teams_at_cutoff": list(set(tied_teams_at_cutoff)),
        "advancing_team_ids": advancing_team_ids,
        "eliminated_team_ids": eliminated_team_ids
    }
